### PR TITLE
crimson/net: postpone the close() using yield()

### PR DESCRIPTION
**Pull request by tchaikov**
otherwise we'd erase an element in a container when we are still
iterating through it.

Fixes: https://tracker.ceph.com/issues/56589
Signed-off-by: Kefu Chai <tchaikov@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/net/Protocol.h [hunk 1]
```diff
@@ -34,11 +35,18 @@ class Protocol {
...
-    // it can happen if close_clean() is called inside Dispatcher::ms_handle_reset()
-    // which will otherwise result in deadlock
-    assert(close_ready.valid());
-    return close_ready.get_future();
+    // yield() so that close(dispatch_reset) can be called *after*
+    // close_clean() is applied to all connections in a container using
+    // seastar::parallel_for_each(). otherwise, we could erase a connection in
+    // the container when seastar::parallel_for_each() is still iterating in
+    // it. that'd lead to a segfault.
+    return seastar::yield().then([this, dispatch_reset] {
```
Comment (liu-chunmei) [inline src/crimson/net/Protocol.h:43]: @tchaikov so the yield will cause this continuation is  scheduled after the other tasks finished in the same schedule group, right?

**Comment by rosinL** [comment]
Does it depends on  the timing `yeild` scheduled back? It is more diffcult to use `parallel_for_each` like this way. Maybe we need two kinds of  `parallel_for_each` for  move-olny iterator and copyable iterator?

**Comment by cyx1231st** [comment]
> Does it depends on the timing yield scheduled back?

I think not, as long as yield can put the task into a different continuation to be scheduled later, it should be fine.

**Review by cyx1231st — APPROVED** [review]

File: src/crimson/net/Protocol.h (hunk 1 above)
Comment (tchaikov) [reply src/crimson/net/Protocol.h:43]: @liu-chunmei kinda. as the terminology of `scheduling_group` has its specific meaning in Seastar, i'd prefer describe the solution as: call `close()` after `seastar::parallel_for_each()`. as without this change, both the futurized continuations (i.e., the `close()` calls) created by `seastar::parallel_for_each()` and its body are executed in the same continutation if `SEASTAR_DEBUG` is not defined. after this change, the body of `seastar::parallel_for_each()` is executed first. only after all the futures are collected, the continuation pending on `yield()` are scheduled.

**Pull request merged**